sql/src/main/java/io/crate/execution/jobs/DistResultRXTask.java

```diff
@@ -24,299 +24,62 @@
     private final int numBuckets;
-    private final Set<Integer> buckets;
-    private final Set<Integer> exhausted;
-    private final PagingIterator<Integer, Row> pagingIterator;
-    private final Map<Integer, PageResultListener> listenersByBucketIdx;
-    private final Map<Integer, Bucket> bucketsByIdx;
-    private final RowConsumer consumer;
-    private final BatchPagingIterator<Integer> batchPagingIterator;
-
-    private Throwable lastThrowable = null;
-    private volatile boolean receivingFirstPage = true;
+    private final BucketReceiver bucketReceiver;
 
     public DistResultRXTask(Logger logger,
-                            String nodeName,
                             int id,
                             String name,
-                            Executor executor,
-                            RowConsumer rowConsumer,
-                            PagingIterator<Integer, Row> pagingIterator,
-                            Streamer<?>[] streamers,
+                            BucketReceiver bucketReceiver,
                             RamAccountingContext ramAccountingContext,
                             int numBuckets) {
         super(id, logger);
-        this.nodeName = nodeName;
         this.name = name;
-        this.executor = executor;
-        this.streamers = streamers;
         this.ramAccountingContext = ramAccountingContext;
         this.numBuckets = numBuckets;
-        this.buckets = new HashSet<>(numBuckets);
-        traceEnabled = logger.isTraceEnabled();
-        this.exhausted = new HashSet<>(numBuckets);
-        this.pagingIterator = pagingIterator;
-        this.bucketsByIdx = new HashMap<>(numBuckets);
-        this.listenersByBucketIdx = new HashMap<>(numBuckets);
-        batchPagingIterator = new BatchPagingIterator<>(
-            pagingIterator,
-            this::fetchMore,
-            this::allUpstreamsExhausted,
-            () -> releaseListenersAndCloseContext(null)
-        );
-        this.consumer = rowConsumer;
-    }
-
-    private void releaseListenersAndCloseContext(@Nullable Throwable throwable) {
-        synchronized (buckets) {
-            for (PageResultListener resultListener : listenersByBucketIdx.values()) {
-                resultListener.needMore(false);
-            }
-            listenersByBucketIdx.clear();
-        }
-        close(throwable);
-    }
-
-    private boolean allUpstreamsExhausted() {
-        return exhausted.size() == numBuckets;
-    }
+        this.bucketReceiver = bucketReceiver;
 
-    @Override
-    public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultListener pageResultListener) {
-        synchronized (buckets) {
-            buckets.add(bucketIdx);
-            if (lastThrowable == null) {
-                listenersByBucketIdx.put(bucketIdx, pageResultListener);
+        this.bucketReceiver.completionFuture().whenComplete((result, ex) -> {
+            if (ex == null) {
+                releaseListenersAndCloseContext(null);
```

Comment:
the null check is not needed if the other branch is only true when using `instanceof`.

Reply:
of course :) thanks, good catch!